pkg/minikube/constants/constants.go

```diff
@@ -62,6 +67,45 @@ func getMinishiftHomeDir() string {
 	homeEnv, ok := os.LookupEnv(MiniShiftHomeEnv)
 	if ok {
 		return homeEnv
+	} else if IsProfileExists() {
+		return filepath.Join(util.HomeDir(), ".minishift", GetMachineName())
+	} else {
+		return filepath.Join(util.HomeDir(), ".minishift")
+	}
+}
+
+//Checks viper or allinstanceconfg to say if profile is defined.
+func IsProfileExists() bool {
+	if viper.GetString("profile") != "" {
+		return true
+	} else if minishiftConfig.AllInstancesConfig != nil {
+		if profileActions.GetActiveProfile() != "" {
+			return true
+		}
+	}
+	return false
+}
+
+//Returns the machines after taking in to account if --profile or
+//profile information available in allinstance config
+func GetMachineName() string {
+	activeProfile := profileActions.GetActiveProfile()
+	if viper.GetString("profile") != "" {
+		return viper.GetString("profile")
+	} else if activeProfile != "" {
+		return activeProfile
+	} else {
+		return DefaultMachineName
+	}
+}
+
+//Returns Minishift instance path. Its considers the profile if defined and used
+func GetMinipath() string {
+	homeEnv, ok := os.LookupEnv(MiniShiftHomeEnv)
+	if ok {
+		return homeEnv
+	} else if IsProfileExists() {
+		return filepath.Join(util.HomeDir(), ".minishift", GetMachineName())
```

Comment:
I don't think this is correct. MINISHIFT_HOME determines the home directory for all the machines. It effectively changes the location of the default ~/.minishift. This means, if it is set we need to take it into accound when building the path to the profile. In your case you are returning from this method if it is set and if not you are just assuming ~/.minishift.

I really think we need unit tests here. They will help thinking this all trough and they also serve as living documentation of how we expect this to work.